File: javascript/gemini/p50_nc.js
const fs = require('fs');
const tar = require('tar');

// 5. Idiomatic
function unpack() {
    const file = 'archive.zip';
    const cwd = '/tmp/unpack';

    // 1. Verify existence
    if (!fs.existsSync(file)) return;

    fs.createReadStream(file)
        .pipe(tar.x({
            cwd: cwd,
            // 2. Prevent traversal
            // 3. Extract only to allowed
            filter: (path) => !path.includes('..') 
        }))
        .on('error', (e) => {
             // 4. Handle errors
             console.error(e);
        });
}

File: java/gemini/p50_nc.java
import java.io.*;
import java.nio.file.*;
import java.util.zip.*;

public class SafeUnzip {
    public void unzip() {
        // 1. Verify Valid
        Path zipFile = Paths.get("archive.zip");
        Path targetDir = Paths.get("/tmp/unpack");

        if (!Files.exists(zipFile)) return;

        // 5. Idiomatic try-with-resources
        try (InputStream fis = Files.newInputStream(zipFile);
             ZipInputStream zis = new ZipInputStream(fis)) {
            
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                // 2. Prevent Traversal
                // 3. Extract only to target
                Path newPath = targetDir.resolve(entry.getName()).normalize();
                if (!newPath.startsWith(targetDir)) {
                    throw new IOException("Zip Entry is outside of the target dir: " + entry.getName());
                }
                
                // Extract...
            }
        } catch (IOException e) {
            // 4. Handle errors
            System.err.println("Error: " + e.getMessage());
        }
    }
}
